File: backend/app/execution/capabilities/calculation_mode.py
"""calculationMode — set workbook calculation mode (manual / automatic / automaticExceptTables)."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry


# Excel COM CalculationMode enum:
#   -4135 = xlCalculationAutomatic
#   -4105 = xlCalculationSemiautomatic (AutomaticExceptTables)
#   -4135 is wrong in some docs; the actual values are:
#   xlCalculationAutomatic = -4105, xlCalculationManual = -4135,
#   xlCalculationSemiautomatic = 2. Use xlwings constants where available.
_MODE_MAP = {
    "automatic": -4105,               # xlCalculationAutomatic
    "automaticexcepttables": 2,       # xlCalculationSemiautomatic
    "manual": -4135,                  # xlCalculationManual
}


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    mode = (params.get("mode") or "").strip().lower()
    if mode not in _MODE_MAP:
        return {
            "status": "error",
            "message": f"Unknown calc mode: {params.get('mode')!r}. Expected one of: manual, automatic, automaticExceptTables.",
        }

    if ctx.dry_run:
        return {"status": "preview", "message": f"Would set calculation mode to {mode}."}

    try:
        ctx.workbook_handle.app.api.Calculation = _MODE_MAP[mode]
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Setting calc mode failed: {exc}", "error": str(exc)}

    return {"status": "success", "message": f"Calculation mode set to {mode}."}


registry.register("calculationMode", handler, mutates=False)

File: frontend/src/engine/capabilities/calculationMode.ts
/**
 * calculationMode — set workbook calculation mode.
 *
 * Office.js: workbook.application.calculationMode (ExcelApi 1.8+).
 * Values: "Automatic" | "AutomaticExceptTables" | "Manual".
 *
 * Useful as a bracket around large bulk-writes to avoid recalc thrashing.
 */

import { CapabilityMeta, CalculationModeParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";

const meta: CapabilityMeta = {
  action: "calculationMode",
  description: "Set workbook calculation mode (manual / automatic / automaticExceptTables)",
  mutates: false,
  affectsFormatting: false,
  requiresApiSet: "ExcelApi 1.8",
};

async function handler(
  context: Excel.RequestContext,
  params: CalculationModeParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { mode } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would set calculation mode to ${mode}` };
  }

  const modeMap: Record<string, Excel.CalculationMode> = {
    manual:                 Excel.CalculationMode.manual,
    automatic:              Excel.CalculationMode.automatic,
    automaticExceptTables:  Excel.CalculationMode.automaticExceptTables,
  };
  const target = modeMap[mode];
  if (!target) {
    return { stepId: "", status: "error", message: `Unknown calculation mode: ${mode}` };
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (context.workbook as any).application.calculationMode = target;
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Calculation mode set to ${mode}.`,
  };
}

// Fallback: pre-1.8 has no programmatic calc mode
async function fallback(
  _context: Excel.RequestContext,
  params: CalculationModeParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  if (options.dryRun) {
    return { stepId: "", status: "success", message: "Would skip calc mode (legacy fallback)." };
  }
  return {
    stepId: "",
    status: "success",
    message:
      `Calculation mode request (${params.mode}) skipped — Application.calculationMode requires ExcelApi 1.8+. ` +
      `Set manually via Formulas > Calculation Options (legacy-Excel fallback).`,
  };
}

registry.register(meta, handler as any, fallback as any);
export { meta };
